### PR TITLE
Android 14 Runtime registered broadcast receivers export behavior 

### LINKED ISSUE #1857: Android 14 Runtime-registered broadcasts receivers must specify export behavior
### Description

Android 14 Runtime-registered broadcast receivers must specify export behavior.

SdlRouterService and LockScreenManager both have context-registered broadcast receivers that need to specify their export behavior.

### PR DESCRIPTION
**Pull request by JulianKast**
Fixes #1857

This PR is **[ready]** for review.

### Risk
This PR makes **[no]** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).
- [x] I have tested Android

#### Unit Tests
No additional unit test were added

#### Core Tests
Note used an integration branch with All Android 14 PRs to test.
Build variants on Android 14 test plan card on trello.

All Test are designed to trigger BroadcastReceivers that I added the export flag behavior.

Test 1:
Install app A
Connect to TDK via Bluetooth
Enable moving on TDK
Observe:  LockScreen in enabled.

Test 2
Add LockScreenConfig to SdlManager builder with enableDismissGesture to true.
Install Hello_Sdl with TCP build variant.
Connect to manticore
Enable DD
Dismiss LockScreen.
Observe: that LockScreen dismisses.

Test 3.
Install App A
Connect to tdk via Bluetooth
Observe App A connect to tdk
Install App B (Note, Before you install App B, unplug the phone from the computer and switch the build variant before deploying to the phone otherwise Android Studio will kill App A as it deploys App B)
Observe: App B connects to tdk.

Core version / branch / commit hash / module tested against: Sync 3 and manticore
HMI name / version / branch / commit hash / module tested against: Sync 3 and manticore

### Summary
This PR updates the gradle files for SDL and Hello SDL to target API level 34, as well as specifies export behavior for runtime-registered broadcasts receivers. 

If a BroadcastReceiver only receives System broadcast, it is not necessary to add the export flag to when we register.

##### Bug Fixes
* Tragets Android 14 and updates context registered broadcast receiver behavior for API 34

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Review by joeygrover — CHANGES_REQUESTED** [review]
Couple of places I don't believe the receivers need to be exported. I think it might be helpful to make a method in AndroidTools to handle the version check to make the code more concise (Note: we use the Oreo code because that's when Android added the method signature that includes flags):

```java
    /**
     * A helper method to handle adding flags to registering a run time broadcast receiver.
     *
     * @param context  a context that will be used to register the receiver with
     * @param receiver the receiver that will be registered
     * @param filter   the filter that will be use to filter intents sent to the broadcast receiver
     * @param flags    any flags that should be used to register the receiver. In most cases this
     *                 will be {@link  Context#RECEIVER_NOT_EXPORTED} or
     *                 {@link  Context#RECEIVER_EXPORTED}
     * @see Context#registerReceiver(BroadcastReceiver, IntentFilter)
     * @see Context#registerReceiver(BroadcastReceiver, IntentFilter, int)
     */
    @SuppressLint("UnspecifiedRegisterReceiverFlag")
    public static void registerReceiver(Context context, BroadcastReceiver receiver, IntentFilter filter, int flags) {
        if (context != null && receiver != null && filter != null) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                context.registerReceiver(receiver, filter, flags);
            } else {
                context.registerReceiver(receiver, filter);
            }
        }
    }
```

I also found two more instances of where we register a receiver that could use the new flag for correctness:

- [TransportManager line 415](https://github.com/smartdevicelink/sdl_java_suite/blob/3e714213b835b2b1f60662e62d397d7db62b2be2/android/sdl_android/src/main/java/com/smartdevicelink/transport/TransportManager.java#L415) - Exported
- [MediaStreamingStatus line 264](https://github.com/smartdevicelink/sdl_java_suite/blob/3e714213b835b2b1f60662e62d397d7db62b2be2/android/sdl_android/src/main/java/com/smartdevicelink/util/MediaStreamingStatus.java#L264) - Exported

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/SDLLockScreenActivity.java [hunk 1]
```diff
@@ -104,9 +105,12 @@ protected void onCreate(Bundle savedInstanceState) {
         IntentFilter lockscreenFilter = new IntentFilter();
         lockscreenFilter.addAction(CLOSE_LOCK_SCREEN_ACTION);
         lockscreenFilter.addAction(LOCKSCREEN_DEVICE_LOGO_DOWNLOADED);
-
         // register broadcast receivers
-        registerReceiver(lockScreenBroadcastReceiver, lockscreenFilter);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+            registerReceiver(lockScreenBroadcastReceiver, lockscreenFilter, RECEIVER_EXPORTED);
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/SDLLockScreenActivity.java]: The lockscreen won't dismiss if it's not exported

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/LockScreenManager.java [hunk 2]
```diff
@@ -332,9 +333,12 @@ private void launchLockScreenActivity() {
             // pass in icon, background color, and custom view
             if (lockScreenEnabled && isApplicationForegrounded && context.get() != null) {
                 if (isLockscreenDismissible && !lockscreenDismissReceiverRegistered) {
-                    context.get().registerReceiver(mLockscreenDismissedReceiver, new IntentFilter(SDLLockScreenActivity.KEY_LOCKSCREEN_DISMISSED));
+                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
+                        context.get().registerReceiver(mLockscreenDismissedReceiver, new IntentFilter(SDLLockScreenActivity.KEY_LOCKSCREEN_DISMISSED), Context.RECEIVER_EXPORTED);
```
[the remark this answers is not in the document]
Comment (JulianKast) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/LockScreenManager.java]: This receiver does not work unless it is exported.

[the remark this answers is not in the document]
Comment (joeygrover) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/LockScreenManager.java]: The issue is that the original intent does not have a package attached to it. So do a find usage on the intents received by this receiver, and where they are created the following line needs to get added before they are sent via broadcast and then setting these to not exported works correctly:

```java
intent.setPackage(context.get().getPackageName())
```

File: android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/SDLLockScreenActivity.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (joeygrover) [reply android/sdl_android/src/main/java/com/smartdevicelink/managers/lockscreen/SDLLockScreenActivity.java]: Same comment as above.

**Review by joeygrover — APPROVED** [review]

**Pull request merged**